### PR TITLE
Fix #29 - Updates backend from sendnn_decoder to sendnn

### LINKED ISSUE #29: Deprecated backend `sendnn_decoder`,
When using the `model.compile(backend="sendnn_decoder")` the backend shows the following warning:

```shell
You're using a deprecated backend. Please use sendnn in conjunction with warmup_mode
```
This is because we are not warming up using `warmup_mode`  with `backend=“sendnn”`, it should be done as following:

```python
model.forward = torch.compile(model.forward, backend = "sendnn", dynamic=False)

# warmup
with torch_sendnn.warmup_mode():
     generate(model, ...)
```

It is currently being used in the following locations:

https://github.com/foundation-model-stack/aiu-fms-testing-utils/blob/6e7e360fcb5c670ece877e25938e751dcdbf55ee/scripts/inference.py#L467

https://github.com/foundation-model-stack/aiu-fms-testing-utils/blob/6e7e360fcb5c670ece877e25938e751dcdbf55ee/scripts/validation.py#L465

https://github.com/foundation-model-stack/aiu-fms-testing-utils/blob/6e7e360fcb5c670ece877e25938e751dcdbf55ee/tests/models/test_decoders.py#L336

### PR DESCRIPTION
**Pull request by flaviabeo**
Fixes issue #29

### CODE REVIEW

File: aiu_fms_testing_utils/utils/__init__.py [hunk 1]
```diff
@@ -12,6 +12,9 @@
 
 def warmup_model(model: nn.Module, input_ids: torch.Tensor, max_new_tokens: int, **padding_kwargs):
     from torch_sendnn import torch_sendnn
+    import torch_sendnn
+
+    torch_sendnn.warmup_mode()
```
Comment (ani300) [inline aiu_fms_testing_utils/utils/__init__.py:17]: this is not how this works. warmup_mode is a python context, so the code should look like this:

```python
    dprint("AIU warmup")
    pt_compile_model_time = time.time()
    extra_kwargs = {**padding_kwargs, "only_last_token": True}
    with torch_sendnn.warmup_mode():
        generate(model, input_ids, max_new_tokens=max_new_tokens, max_seq_len=model.config.max_expected_seq_len, use_cache=True, do_sample=False, contiguous_cache=True, extra_kwargs=extra_kwargs)
    pt_compile_model_time = time.time() - pt_compile_model_time
    dprint(f"PT compile complete, took {pt_compile_model_time:.3f}s")

    # All the code in lines 25-29 is no longer needed (update_lazyhandle is now implicit)
```

**Review by ani300 — CHANGES_REQUESTED** [review]
left two comments on how to properly use `warmup_mode`

File: scripts/inference.py [hunk 2]
```diff
@@ -645,6 +645,9 @@ def infer(use_cache, do_sample, warmup):
         eos_token_id = tokenizer.eos_token_id
     else:
         eos_token_id = None
+        
+    if warmup:
+        torch_sendnn.warmup_mode()
```
[the remark this answers is not in the document]
Comment (flaviabeo) [reply scripts/inference.py:650]: thank you!

**Comment by ani300** [comment]
bot:test
TEST_FILE=test_decoders.py MODEL_ID=ibm-granite/granite-3.2-8b-instruct BATCH_SIZE=1,8 SEQUENCE_LENGTH=64 USE_TINY_MODEL=1

**Comment by ani300** [comment]
@flaviabeo have you tested all the scripts/tests work? I've been busy with other things and haven't been able to test them. If you can check that before/after the changes the scripts/tests pass with the same model, I can merge this

**Comment by flaviabeo** [comment]
> @flaviabeo have you tested all the scripts/tests work? I've been busy with other things and haven't been able to test them. If you can check that before/after the changes the scripts/tests pass with the same model, I can merge this

@ani300 I have run inference.py couple times, yes - but I've just triggered another one to confirm. Will ping you as soon as I get the results

**Review by ani300 — APPROVED** [review]
changes now lgtm! waiting for final confirmation both inference.py and test_decoders works before merging

**Comment by flaviabeo** [comment]
@ani300 @JRosenkranz this afternoon latest results:
- Inference:
```bash
Starting to run inference granite-20b-code-cobol-v1
/tmp/foundation-model-stack/fms/triton/pytorch_ops.py:103: FutureWarning: `torch.library.impl_abstract` was renamed to `torch.library.register_fake`. Please use that instead; we will remove `torch.library.impl_abstract` in a future version of PyTorch.
@torch.library.impl_abstract("moe::moe_mm")
[ 0/ 1]: Namespace(device_type='aiu', architecture='hf_pretrained', variant=None, model_path='/ibm-dmf/models/watsonx/shared/granite-20b-code-cobol-v1/20240603', model_source=None, quantization=None, int8_weight_per_channel=False, int8_activ_quant_type='per_token', int8_smoothquant=False, tokenizer='/ibm-dmf/models/watsonx/shared/granite-20b-code-cobol-v1/20240603', no_use_cache=True, unfuse_weights=True, default_dtype='fp16', compile=False, compile_mode='default', compile_backend='inductor', compile_dynamic=True, deterministic=False, distributed=False, batch_size=1, max_prompt_length=None, min_pad_length=0, fixed_prompt_length=128, max_new_tokens=16, no_early_termination=False, prompt_type='code', prompt_path='', output_path='/mnt/aiu-models-en-shared/granite-20b-code-cobol-v1/output_fms_22_apr_25_inference_granite_20b_code_cobol_v1', timing='per-token', iters=1, verbose=0)
[ 0/ 1]: Sentient AIU: Enabled
[ 0/ 1]: NOTICE: Adjusting torch._dynamo.config.accumulated_cache_size_limit from 256 to 320 to accomodate prompt size of 128 and decode tokens of 16
[ 0/ 1]: NOTICE: Adjusting torch._dynamo.config.cache_size_limit from 8 to 320 to accomodate prompt size of 128 and decode tokens of 16
[ 0/ 1]: loading model
[ 0/ 1]: ============================================================
[ 0/ 1]: model_path=/ibm-dmf/models/watsonx/shared/granite-20b-code-cobol-v1/20240603
[ 0/ 1]: linear_config={'linear_type': 'torch_linear'}
[ 0/ 1]: fused_weights=False
[ 0/ 1]: data_type=torch.float16
[ 0/ 1]: ============================================================
[ 0/ 1]: loading complete, took 4660.552s
[ 0/ 1]: generating output
[ 0/ 1]: use_cache True;; do_sample False
[ 0/ 1]: ==================
[ 0/ 1]: First-token latency: 2721880.185 ms
[ 0/ 1]: Average next-token latency: 21606.389 ms
[ 0/ 1]: Average next-token latency (including first token): 190373.501 ms
[ 0/ 1]: Max next-token latency: 22834.018 ms (token #2)
[ 0/ 1]: Min next-token latency: 21313.109 ms (token #6)
[ 0/ 1]: Std deviation of next-token latencies: 492.868 ms
[ 0/ 1]: Per-token timing information: 2721880.185, 22834.018, 21393.875, 21359.256, 21326.193, 21313.109, 21320.962, 21437.502, 21322.014, 21319.768, 21394.717, 21422.559, 21404.304, 21356.324, 22480.031, 22411.198 ms
[ 0/ 1]: [INST] Write code to solve the following coding problem that obeys the constraints and passes the example test cases. Please wrap your code answer using ```:
Write a bubble sort function in python.
[/INST]
```python
def bubble_sort(arr):
for i in
Finished running inference
```
- Test decoders 
```bash
[ 0/ 1]: update_lazyhandle complete, took 34.404s
[ 0/ 1]: cpu validation info extracted for validation level 0 and validation level 1 (iter=0)
[ 0/ 1]: aiu validation info extracted for validation level 0
[ 0/ 1]: passed validation level 0, testing validation level 1
[ 0/ 1]: aiu validation info extracted for validation level 1 - iter=0
[ 0/ 1]: cpu validation info extracted for validation level 1 - iter=1
[ 0/ 1]: aiu validation info extracted for validation level 1 - iter=1
[ 0/ 1]: cpu validation info extracted for validation level 1 - iter=2
[ 0/ 1]: aiu validation info extracted for validation level 1 - iter=2
[ 0/ 1]: cpu validation info extracted for validation level 1 - iter=3
[ 0/ 1]: aiu validation info extracted for validation level 1 - iter=3
[ 0/ 1]: cpu validation info extracted for validation level 1 - iter=4
[ 0/ 1]: aiu validation info extracted for validation level 1 - iter=4
[ 0/ 1]: cpu validation info extracted for validation level 1 - iter=5
[ 0/ 1]: aiu validation info extracted for validation level 1 - iter=5
[ 0/ 1]: cpu validation info extracted for validation level 1 - iter=6
[ 0/ 1]: aiu validation info extracted for validation level 1 - iter=6
[ 0/ 1]: cpu validation info extracted for validation level 1 - iter=7
[ 0/ 1]: aiu validation info extracted for validation level 1 - iter=7
------------------------------ Captured log call -------------------------------
============================= fixture duration top =============================
total name num avg min
0:00:00.000100 grand total 5 0:00:00.000010 0:00:00.000009
============================ test call duration top ============================
total name num avg min
0:02:41.566818 test_common_shapes 1 0:02:41.566818 0:02:41.566818
0:02:41.566818 grand total 1 0:02:41.566818 0:02:41.566818
=========================== test setup duration top ============================
total name num avg min
0:00:00.000367 grand total 1 0:00:00.000367 0:00:00.000367
========================== test teardown duration top ==========================
total name num avg min
0:00:00.000494 grand total 1 0:00:00.000494 0:00:00.000494
=========================== short test summary info ============================
FAILED tests/models/test_decoders.py::test_common_shapes[/ibm-dmf/models/watsonx/shared/granite-20b-code-cobol-v1/20240603-1-64-128]
=================== 1 failed, 1 warning in 162.79s (0:02:42) ===================
Finished running pytests
```

File: scripts/inference.py [hunk 3]
```diff
@@ -688,18 +689,13 @@ def infer(use_cache, do_sample, warmup):
 if args.compile:
     dprint(f"compilation warmup")
     pt_compile_model_time = time.time()
-    for sample, cache in itertools.product(do_sample, use_cache):
-        infer(cache, sample, True)
+    warmup_context = torch_sendnn.warmup_mode() if is_aiu_backend else contextlib.nullcontext()
```
Comment (JRosenkranz) [inline scripts/inference.py:692]: could we re-use `warmup_model` here?

Comment (JRosenkranz) [inline scripts/inference.py:692]: could we re-use `warmup_model` here?

File: scripts/inference.py [hunk 4]
```diff
@@ -688,18 +689,13 @@ def infer(use_cache, do_sample, warmup):
...
+    with warmup_context:
```
Comment (JRosenkranz) [inline scripts/inference.py:693]: could we re-use `warmup_model` here?

File: scripts/inference.py (hunk 3 above)
Comment (flaviabeo) [reply scripts/inference.py:692]: I'll try replacing this here

Comment (flaviabeo) [reply scripts/inference.py:692]: @JRosenkranz just finished running a test, and reusing the `warmup_model` in here and it worked fine!

**Review by ani300 — APPROVED** [review]
lgtm now!

File: scripts/inference.py [hunk 5]
```diff
@@ -691,14 +692,21 @@ def infer(use_cache, do_sample, warmup):
...
-    if args.device_type == "aiu":  # only run warmup for AIU, no need for senulator
-        aiu_warmup_time = time.time()
-        for sample, cache in itertools.product(do_sample, use_cache):
-            infer(cache, sample, True)
-        aiu_warmup_time = time.time() - aiu_warmup_time
-        dprint(f"AIU warmup complete, took {aiu_warmup_time:.3f}s")
+    dprint(f"compilation warmup")
+    import torch_sendnn
+    pt_compile_model_time = time.time()
+    warmup_context = torch_sendnn.warmup_mode() if is_aiu_backend else contextlib.nullcontext()
```
Comment (JRosenkranz) [inline scripts/inference.py:698]: I don't believe we need this `warmup_context` here anymore as warmup_model is already using `warmup_mode` context.

**Comment by JRosenkranz** [comment]
bot:test
TEST_FILE=test_decoders.py MODEL_ID=ibm-granite/granite-3.2-8b-instruct BATCH_SIZE=1,8 SEQUENCE_LENGTH=64 USE_TINY_MODEL=0

**Comment by JRosenkranz** [comment]
bot:test
TEST_FILE=test_decoders.py MODEL_ID=ibm-granite/granite-3.2-8b-instruct BATCH_SIZE=1,8 SEQUENCE_LENGTH=64 USE_TINY_MODEL=0

File: scripts/inference.py [hunk 6]
```diff
@@ -693,18 +693,15 @@ def infer(use_cache, do_sample, warmup):
...
-            aiu_warmup_time = time.time()
-            warmup_model(model=model, input_ids=ids, max_new_tokens=args.max_new_tokens)
-            aiu_warmup_time = time.time() - aiu_warmup_time
-            dprint(f"AIU warmup complete, took {aiu_warmup_time:.3f}s")
-        else:
-            for sample, cache in itertools.product(do_sample, use_cache):
-                infer(cache, sample, True)
+    if args.device_type == "aiu":  # only run warmup for AIU, no need for senulator
+        aiu_warmup_time = time.time()
+        warmup_model(model=model, input_ids=ids, max_new_tokens=args.max_new_tokens)
```
Comment (JRosenkranz) [inline scripts/inference.py:699]: This warmup will need to include `args.compile_dynamic_sendnn` and `**extra_generation_kwargs`.

I think the below should look like this:

```python
if args.device_type == "aiu":  # only run warmup for AIU, no need for senulator
        warmup_model(model, ids, args.max_new_tokens, args.compile_dynamic_sendnn, **extra_generation_kwargs)

        aiu_warmup_time = time.time()
        for sample, cache in itertools.product(do_sample, use_cache):
            infer(cache, sample, True)
        aiu_warmup_time = time.time() - aiu_warmup_time
        dprint(f"AIU warmup complete, took {aiu_warmup_time:.3f}s")
```

**Comment by JRosenkranz** [comment]
bot:test
TEST_FILE=test_decoders.py MODEL_ID=ibm-granite/granite-3.2-8b-instruct BATCH_SIZE=1 SEQUENCE_LENGTH=64 USE_TINY_MODEL=0
